protocol-messages.go

```diff
@@ -0,0 +1,380 @@
+package mls
+
+import (
+	"fmt"
+	"github.com/bifurcation/mint/syntax"
+)
+
+////// ClientInitKey
+
+// enum {
+//	invalid(0),
+//	supported_versions(1),
+//	supported_ciphersuites(2),
+//	expiration(3),
+//	(65535)
+// } ExtensionType;
+
+// struct {
+//	ExtensionType extension_type;
+//	opaque extension_data<0..2^16-1>;
+// } Extension;
+
+type ExtensionType uint16
+type Extension struct {
+	ExtensionType ExtensionType
+	ExtensionData []byte `tls:"head=2"`
+}
+
+// struct {
+//	ProtocolVersion supported_version;
+//	opaque client_init_key_id<0..255>;
+//	CipherSuite cipher_suite;
+//	HPKEPublicKey init_key;
+//	Credential credential;
+//	Extension extensions<0..2^16-1>;
+//	opaque signature<0..2^16-1>;
+// } ClientInitKey;
+type ClientInitKey struct {
+	ProtocolVersion uint8
+	CipherSuite     CipherSuite
+	HPKEPublicKey   []byte `tls:"head=2"`
+	Credential      Credential
+	Extensions      []Extension `tls:"head=2"`
+	Signature       []byte      `tls:"head=2"`
+}
+
+//// Proposal
+type ProposalType uint8
+
+const (
+	invalid = iota
+	add
+	update
+	remove
+)
+
+type AddProposal struct {
+	ClientInitKey ClientInitKey
+}
+
+type UpdateProposal struct {
+	// this is a HPKEPublicKey (for this structure purposes it is opaque)
+	LeafKey []byte `tls:"head=2"`
+}
+
+type RemoveProposal struct {
+	Removed uint32
+}
+
+type Proposal struct {
+	Type   ProposalType
+	Add    *AddProposal
+	Update *UpdateProposal
+	Remove *RemoveProposal
+}
+
+func (p Proposal) MarshalTLS() ([]byte, error) {
+	s := &Stream{}
+	err := s.Write(p.Type)
+	if err != nil {
+		fmt.Errorf("mls.proposal: Marshal failed for ProposalType: %v", err)
+	}
+	syntax.Marshal(p.Type)
+	switch p.Type {
+	case add:
+		err = s.Write(p.Add)
+	case update:
+		err = s.Write(p.Update)
+	case remove:
+		err = s.Write(p.Remove)
+	default:
+		return nil, fmt.Errorf("mls.proposal: ProposalType type not allowed: %v", err)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("mls.proposal: Marshal failed: %v", err)
+	}
+	return s.Data(), nil
+}
+
+func (p *Proposal) UnmarshalTLS(data []byte) (int, error) {
+	s := &Stream{}
+	var ptype ProposalType
+	read, err := s.Read(data, &ptype)
+	if err != nil {
+		fmt.Errorf("mls.mlsproposal: Unmarshal failed for proposal type")
+		return 0, err
+	}
+	p.Type = ptype
+	switch p.Type {
+	case add:
+		var inner AddProposal
+		read, err := s.Read(data[read:], &inner)
+		if err != nil {
+			return 0, err
+		}
+		p.Add = &inner
+		return read, nil
+	case update:
+		var inner UpdateProposal
+		read, err := s.Read(data[read:], &inner)
+		if err != nil {
+			return 0, err
+		}
+		p.Update = &inner
+		return read, nil
+	case remove:
+		var inner RemoveProposal
+		read, err := s.Read(data[read:], &inner)
+		if err != nil {
+			return 0, err
+		}
+		p.Remove = &inner
+		return read, nil
+	default:
+		return 0, fmt.Errorf("mls.proposal: ProposalType type not allowed")
+	}
+}
+
+//// commit
+type ProposalId struct {
```

Comment:
Note: Spec and `mlspp` differ on this struct.  To make things work, you'll probably want to use the `mlspp` version.